### ISSUE TITLE
Make busy waiting time in DDP (kSynchronizeBusyWaitMillis) a configurable environment variable

### ISSUE THREAD

**Issue by leitian** [pytorch/pytorch#43510]
## 🚀 Feature

Make busy waiting time in DDP (kSynchronizeBusyWaitMillis) a configurable environment variable, so users can customize the checking interval via environment variable.

https://github.com/pytorch/pytorch/blob/6e1127ea3f64034329062ca9ad6c60a553809566/torch/lib/c10d/ProcessGroupNCCL.cpp#L232

## Motivation

We spot a potential performance issue in the case of NCCL_BLOCKING_WAIT being turned on and model with many SyncBatchNorm operators. When NCCL_BLOCKING_WAIT is turned on, GPU is kept idle after NCCL kernel finishes the execution since SyncBatchNorm checks the completion of NCCL every 10 milli-seconds (the default value of kSynchronizeBusyWaitMillis). Making it configurable allows user to easily tune the checking interval for performance improvement in user's scenarios.

## Pitch

1. Create a environment variable for the checking interval, Say "NCCL_POLLING_INTERVAL_MICROSECONDS", and let DDP load the checking interval from the environment variable;
2. If NCCL_POLLING_INTERVAL_MICROSECONDS is not set by user, we will use 10000, which is the default value of kSynchronizeBusyWaitMillis currently.

**Comment by pritamdamania87** [comment]
I think after https://github.com/pytorch/pytorch/pull/41052 lands, we can completely remove the busy polling and just replace it with `cudaEventSynchronize`. @osalpekar Do you see any potential issues with this approach? My understanding is that the timeouts will be handled by a separate thread and once that thread times out the operation, `cudaEventSynchronize` should return indicating the work has finished (although unsuccessfully)? Once `cudaEventSynchronize` returns, we can check for errors again and throw an appropriate exception?

Could you try this out and see if it works as expected?

**Comment by osalpekar** [comment]
Yes I think we should go with the `cudaEventSynchronize` approach. Once the NCCL Async Error Handling stack lands, I can run some experiments comparing performance of `cudaEventSynchronize` vs. various busy polling durations.
